R/data_merge.R

```diff
@@ -0,0 +1,328 @@
+#'
+#' # add ID common column
+#' x$id <- 1:5
+#' y$id <- 3:7
+#'
+#' # left-join, no matching rows (because columns "id" and "disp" are used)
+#' # new variables get all NA values
+#' data_merge(x, y)
+#'
+#' # one common value in "mpg", so one row from y is copied to x
+#' data_merge(x, y, by = "mpg")
+#'
+#' # only keep rows with matching values in by-column
+#' data_merge(x, y, join = "semi", by = "mpg")
+#'
+#' # only keep rows with non-matching values in by-column
+#' data_merge(x, y, join = "anti", by = "mpg")
+#'
+#' # merge list of data frames. can be of different rows
+#' x <- mtcars[1:5, 1:3]
+#' y <- mtcars[28:31, 3:5]
+#' z <- mtcars[11:18, c(1, 3:4, 6:8)]
+#' x$id <- 1:5
+#' y$id <- 4:7
+#' z$id <- 3:10
+#' data_merge(list(x, y, z), join = "bind", by = "id", id = "source")
+#'
+#' @export
+data_merge <- function(x, ...) {
+  UseMethod("data_merge")
+}
+
+#' @rdname data_merge
+#' @export
+data_join <- data_merge
+
+#' @rdname data_merge
+#' @export
+data_merge.data.frame <- function(x, y, join = "left", by = NULL, id = NULL, verbose = TRUE, ...) {
+  attr_x <- attributes(x)
+  attr_y <- attributes(y)
+
+  # check join-argument ----------------------
+
+  join <- match.arg(join, choices = c("full", "left", "right", "inner", "semi", "anti", "bind"))
+
+
+  # check id-argument ----------------------
+
+  all_columns <- union(colnames(x), colnames(y))
+
+  if (!is.null(id) && id %in% all_columns) {
+    if (isTRUE(verbose)) {
+      warning(
+        insight::format_message(sprintf("Value of `id` already exists as column name. ID column was renamed to `%s`.", id)),
+        call. = FALSE
+      )
+    }
+    id <- make.unique(c(all_columns, id), sep = "_")[length(all_columns) + 1]
+  }
+
+  if (!is.null(id) && join == "bind") {
+    x[[id]] <- 1
+    y[[id]] <- 2
+  }
+
+
+  # check merge columns ("by"-argument) ----------------------
+
+  if (join != "bind") {
+
+    # we need a value for "by". If not provided, use all shared column names
+    if (is.null(by)) {
+      by <- intersect(colnames(x), colnames(y))
+    }
+
+    # If not all column names specified in "by" are present, yield warning
+    # and use all shared column names
+    if (!all(by %in% colnames(x)) || !all(by %in% colnames(y))) {
+      if (isTRUE(verbose)) {
+        warning(
+          insight::format_message(
+            "Not all columns specified in `by` were found in the data frames.",
+            "Using all columns that are present in both data frames."
+          ),
+          call. = FALSE
+        )
+      }
+      by <- intersect(colnames(x), colnames(y))
+    }
+
+    # if still both data frames have no common columns, do a full join
+    if (!length(by)) {
+      if (isTRUE(verbose)) {
+        warning(
+          insight::format_message(
+            "Found no matching columns in the data frames. Fully merging both data frames now. Note that this can lead to unintended results, because rows in `x` and `y` are possibly duplicated.",
```

Comment:
Can we break this long message down into smaller segments? Just like the other messages in this function?